process/block/metablock.go

```diff
@@ -1854,106 +1779,60 @@ func (mp *metaProcessor) MarshalizedDataToBroadcast(
-			continue
-		}
-
-		if hdr.GetNonce() <= mp.lastNotarizedHdrForShard(currShardId).GetNonce() {
-			continue
-		}
-
-		hashAndBlockMap[currShardId] = append(hashAndBlockMap[currShardId],
-			&hashAndHdr{hdr: hdr, hash: key})
+	lastCrossNotarizedHeaders, err := mp.getLastCrossNotarizedHeaders()
+	if err != nil {
+		return nil, nil, nil, err
 	}
-	mp.mutNotarizedHdrs.RUnlock()
-
-	maxHdrLen := mp.sortShardHeadrs(hashAndBlockMap)
-	return hashAndBlockMap, maxHdrLen, nil
-}
 
-func (mp *metaProcessor) sortShardHeadrs(hashAndBlockMap map[uint32][]*hashAndHdr) int {
 	maxHdrLen := 0
-	for shardId := uint32(0); shardId < mp.shardCoordinator.NumberOfShards(); shardId++ {
-		hdrsForShard := hashAndBlockMap[shardId]
-		if len(hdrsForShard) == 0 {
-			continue
-		}
+	for shardID := uint32(0); shardID < mp.shardCoordinator.NumberOfShards(); shardID++ {
+		hdrsForShard, hdrsHashesForShard := mp.blockTracker.ComputeLongestChain(shardID, lastCrossNotarizedHeaders[shardID])
 
-		sort.Slice(hdrsForShard, func(i, j int) bool {
-			return hdrsForShard[i].hdr.GetNonce() < hdrsForShard[j].hdr.GetNonce()
-		})
+		hdrsMap[shardID] = append(hdrsMap[shardID], hdrsForShard...)
+		hdrsHashesMap[shardID] = append(hdrsHashesMap[shardID], hdrsHashesForShard...)
 
 		tmpHdrLen := len(hdrsForShard)
 		if maxHdrLen < tmpHdrLen {
 			maxHdrLen = tmpHdrLen
 		}
 	}
 
-	return maxHdrLen
-}
-
-func (mp *metaProcessor) getOrderedHdrs(round uint64) ([]*block.Header, [][]byte, map[uint32][]*block.Header, error) {
-	hashAndBlockMap, maxHdrLen, err := mp.getSortedHeadersPerShard(round)
-	if err != nil {
-		return nil, nil, nil, err
-	}
-
-	headersMap := make(map[uint32][]*block.Header)
-	headers := make([]*block.Header, 0)
-	hdrHashes := make([][]byte, 0)
+	orderedHeaders := make([]data.HeaderHandler, 0)
+	orderedHeadersHashes := make([][]byte, 0)
 
+	// copy from map to lists - equality between number of headers per shard
 	for i := 0; i < maxHdrLen; i++ {
-		for shardId := uint32(0); shardId < mp.shardCoordinator.NumberOfShards(); shardId++ {
-			hdrsForShard := hashAndBlockMap[shardId]
+		for shardID := uint32(0); shardID < mp.shardCoordinator.NumberOfShards(); shardID++ {
+			hdrsForShard := hdrsMap[shardID]
+			hdrsHashesForShard := hdrsHashesMap[shardID]
 			if i >= len(hdrsForShard) {
 				continue
 			}
-
-			hdr, ok := hdrsForShard[i].hdr.(*block.Header)
-			if !ok {
+			if hdrsForShard[i].GetRound() > round {
 				continue
 			}
 
-			headers = append(headers, hdr)
-			hdrHashes = append(hdrHashes, hdrsForShard[i].hash)
-			headersMap[shardId] = append(headersMap[shardId], hdr)
+			orderedHeaders = append(orderedHeaders, hdrsForShard[i])
+			orderedHeadersHashes = append(orderedHeadersHashes, hdrsHashesForShard[i])
 		}
 	}
 
-	return headers, hdrHashes, headersMap, nil
+	return orderedHeaders, orderedHeadersHashes, hdrsMap, nil
+}
+
+func (mp *metaProcessor) getTrackedHeaders(round uint64) map[uint32][]data.HeaderHandler {
```

Comment:
i think this could be added to blockTracker.

Reply:
GetTrackedHeaders from block tracker returns a list of sorted headers/hashes for a given shard, you may not want all the tracked headers for all the shards. Here the returned type is a map. I would not make another method GetTrackedHeadersForAllShards only for this case. I think that if some situation requires all the tracked headers it could be solved easily by iterate on each shard.

Reply:
BlockTracker would implement this function as well, with the same input and output. The thing is, all the data and processing inside the method is using only blockTracker data -> it can be done there.

Reply:
Done